File: python/graphics/traffic_light.py
"""Traffic Light — coordinated super-state FSM + preemption + adaptive timing.

Five-phase cycle: NSGreen → NSYellow → AllRed → EWGreen → EWYellow → AllRed.
Light colors derived from phase; invalid combinations (both green) unrepresentable.
Emergency preemption saves and restores phase.
Adaptive green extension: queued vehicles extend green up to a cap.
Pedestrian walk gated to AllRed entry; tracks Walk → Flashing → Inactive.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class Phase(Enum):
    NS_GREEN = "ns_green"
    NS_YELLOW = "ns_yellow"
    ALL_RED = "all_red"
    EW_GREEN = "ew_green"
    EW_YELLOW = "ew_yellow"
    EMERGENCY_ALL_RED = "emergency_all_red"


class LightColor(Enum):
    RED = "red"
    YELLOW = "yellow"
    GREEN = "green"


class PedSignal(Enum):
    DONT_WALK = "dont_walk"
    WALK = "walk"
    FLASHING = "flashing"


class Direction(Enum):
    NS = "ns"
    EW = "ew"


def phase_ns_light(p: Phase) -> LightColor:
    if p == Phase.NS_GREEN: return LightColor.GREEN
    if p == Phase.NS_YELLOW: return LightColor.YELLOW
    return LightColor.RED


def phase_ew_light(p: Phase) -> LightColor:
    if p == Phase.EW_GREEN: return LightColor.GREEN
    if p == Phase.EW_YELLOW: return LightColor.YELLOW
    return LightColor.RED


@dataclass(frozen=True)
class Config:
    ns_green_ms: int = 20_000
    ns_yellow_ms: int = 3_000
    ew_green_ms: int = 20_000
    ew_yellow_ms: int = 3_000
    all_red_ms: int = 2_000
    max_green_extension_ms: int = 10_000
    ped_walk_ms: int = 8_000
    ped_flashing_ms: int = 5_000


class EventKind(Enum):
    PHASE_CHANGED = "phase_changed"
    GREEN_EXTENDED = "green_extended"
    PED_REQUESTED = "ped_requested"
    PED_WALK_STARTED = "ped_walk_started"
    PED_WALK_ENDED = "ped_walk_ended"
    EMERGENCY_STARTED = "emergency_started"
    EMERGENCY_ENDED = "emergency_ended"


@dataclass
class IntersectionEvent:
    kind: EventKind
    at_ms: int
    detail: str = ""


class PedState(Enum):
    INACTIVE = "inactive"
    WALK = "walk"
    FLASHING = "flashing"


@dataclass
class Intersection:
    config: Config
    phase: Phase = Phase.NS_GREEN
    phase_elapsed_ms: int = 0
    phase_duration_ms: int = 0
    cycle_side: Direction = Direction.EW
    ns_queue: int = 0
    ew_queue: int = 0
    ped_requested: bool = False
    _ped_state: PedState = PedState.INACTIVE
    _ped_elapsed_ms: int = 0
    _saved_phase: Optional[tuple[Phase, int, Direction]] = None
    total_elapsed_ms: int = 0
    events: list[IntersectionEvent] = field(default_factory=list)

    def __post_init__(self) -> None:
        self.phase_duration_ms = self._duration_for(self.phase)

    def _duration_for(self, p: Phase) -> int:
        c = self.config
        if p == Phase.NS_GREEN: return c.ns_green_ms
        if p == Phase.NS_YELLOW: return c.ns_yellow_ms
        if p == Phase.EW_GREEN: return c.ew_green_ms
        if p == Phase.EW_YELLOW: return c.ew_yellow_ms
        if p == Phase.ALL_RED: return c.all_red_ms
        if p == Phase.EMERGENCY_ALL_RED: return 10**18
        return 0

    def ns_light(self) -> LightColor: return phase_ns_light(self.phase)
    def ew_light(self) -> LightColor: return phase_ew_light(self.phase)

    def ped_signal(self) -> PedSignal:
        if self._ped_state == PedState.INACTIVE: return PedSignal.DONT_WALK
        if self._ped_state == PedState.WALK: return PedSignal.WALK
        return PedSignal.FLASHING

    def set_ns_queue(self, n: int) -> None: self.ns_queue = n
    def set_ew_queue(self, n: int) -> None: self.ew_queue = n

    def request_ped_crossing(self) -> None:
        if self._ped_state != PedState.INACTIVE: return
        if not self.ped_requested:
            self.ped_requested = True
            self.events.append(IntersectionEvent(
                EventKind.PED_REQUESTED, self.total_elapsed_ms))

    def begin_emergency(self) -> None:
        if self.phase == Phase.EMERGENCY_ALL_RED: return
        self._saved_phase = (self.phase, self.phase_elapsed_ms, self.cycle_side)
        at = self.total_elapsed_ms
        from_phase = self.phase
        self.phase = Phase.EMERGENCY_ALL_RED
        self.phase_elapsed_ms = 0
        self.phase_duration_ms = 10**18
        self._ped_state = PedState.INACTIVE
        self.events.append(IntersectionEvent(EventKind.EMERGENCY_STARTED, at))
        self.events.append(IntersectionEvent(
            EventKind.PHASE_CHANGED, at,
            f"{from_phase.value} -> emergency_all_red"))

    def end_emergency(self) -> None:
        if self.phase != Phase.EMERGENCY_ALL_RED: return
        at = self.total_elapsed_ms
        from_phase = self.phase
        if self._saved_phase is not None:
            p, elapsed, side = self._saved_phase
            self._saved_phase = None
            self.phase = p
            self.phase_elapsed_ms = elapsed
            self.phase_duration_ms = self._duration_for(p)
            self.cycle_side = side
        else:
            self.phase = Phase.ALL_RED
            self.phase_elapsed_ms = 0
            self.phase_duration_ms = self.config.all_red_ms
        self.events.append(IntersectionEvent(EventKind.EMERGENCY_ENDED, at))
        self.events.append(IntersectionEvent(
            EventKind.PHASE_CHANGED, at,
            f"{from_phase.value} -> {self.phase.value}"))

    def _consider_extension(self) -> int:
        direction: Optional[Direction]
        if self.phase == Phase.NS_GREEN: direction = Direction.NS
        elif self.phase == Phase.EW_GREEN: direction = Direction.EW
        else: direction = None
        if direction is None: return 0
        queue = self.ns_queue if direction == Direction.NS else self.ew_queue
        if queue == 0: return 0

        base = self.config.ns_green_ms if direction == Direction.NS else self.config.ew_green_ms
        already = max(0, self.phase_duration_ms - base)
        remaining_cap = max(0, self.config.max_green_extension_ms - already)
        if remaining_cap == 0: return 0
        grant = min(queue * 2000, remaining_cap)
        self.phase_duration_ms += grant
        if grant > 0:
            self.events.append(IntersectionEvent(
                EventKind.GREEN_EXTENDED, self.total_elapsed_ms,
                f"{direction.value} +{grant}ms"))
        return grant

    def _next_phase(self) -> tuple[Phase, Direction]:
        p = self.phase
        if p == Phase.NS_GREEN: return (Phase.NS_YELLOW, self.cycle_side)
        if p == Phase.NS_YELLOW: return (Phase.ALL_RED, Direction.EW)
        if p == Phase.ALL_RED:
            if self.cycle_side == Direction.EW:
                return (Phase.EW_GREEN, Direction.NS)
            return (Phase.NS_GREEN, Direction.EW)
        if p == Phase.EW_GREEN: return (Phase.EW_YELLOW, self.cycle_side)
        if p == Phase.EW_YELLOW: return (Phase.ALL_RED, Direction.NS)
        return (Phase.EMERGENCY_ALL_RED, self.cycle_side)

    def _transition_phase(self) -> bool:
        new_phase, new_side = self._next_phase()
        from_phase = self.phase
        self.phase = new_phase
        self.cycle_side = new_side
        self.phase_elapsed_ms = 0
        self.phase_duration_ms = self._duration_for(new_phase)
        self.events.append(IntersectionEvent(
            EventKind.PHASE_CHANGED, self.total_elapsed_ms,
            f"{from_phase.value} -> {new_phase.value}"))
        if new_phase == Phase.ALL_RED and self.ped_requested:
            self.ped_requested = False
            self._ped_state = PedState.WALK
            self._ped_elapsed_ms = 0
            self.events.append(IntersectionEvent(
                EventKind.PED_WALK_STARTED, self.total_elapsed_ms))
            return True
        return False

    def _advance_pedestrian(self, ms: int) -> None:
        if self._ped_state == PedState.INACTIVE: return
        self._ped_elapsed_ms += ms
        if (self._ped_state == PedState.WALK
            and self._ped_elapsed_ms >= self.config.ped_walk_ms):
            self._ped_state = PedState.FLASHING
            self._ped_elapsed_ms = 0
        elif (self._ped_state == PedState.FLASHING
              and self._ped_elapsed_ms >= self.config.ped_flashing_ms):
            self._ped_state = PedState.INACTIVE
            self._ped_elapsed_ms = 0
            self.events.append(IntersectionEvent(
                EventKind.PED_WALK_ENDED, self.total_elapsed_ms))

    def tick(self, ms: int) -> None:
        self.total_elapsed_ms += ms
        if self.phase == Phase.EMERGENCY_ALL_RED: return

        self.phase_elapsed_ms += ms
        if (self.phase in (Phase.NS_GREEN, Phase.EW_GREEN)
            and self.phase_elapsed_ms >= self.phase_duration_ms):
            self._consider_extension()

        ped_just_started_remaining: Optional[int] = None
        while (self.phase_elapsed_ms >= self.phase_duration_ms
               and self.phase != Phase.EMERGENCY_ALL_RED):
            overflow = self.phase_elapsed_ms - self.phase_duration_ms
            started = self._transition_phase()
            self.phase_elapsed_ms = min(overflow, self.phase_duration_ms)
            if started:
                ped_just_started_remaining = overflow

        if self._ped_state == PedState.INACTIVE:
            ped_advance = 0
        elif ped_just_started_remaining is not None:
            ped_advance = ped_just_started_remaining
        else:
            ped_advance = ms
        self._advance_pedestrian(ped_advance)


def demo() -> None:
    cfg = Config(
        ns_green_ms=1000, ns_yellow_ms=200, ew_green_ms=1000, ew_yellow_ms=200,
        all_red_ms=100, max_green_extension_ms=500,
        ped_walk_ms=500, ped_flashing_ms=300,
    )
    i = Intersection(cfg)
    print(f"initial: phase={i.phase.value} ns={i.ns_light().value} ew={i.ew_light().value}")

    i.tick(1000)
    print(f"after 1000ms: phase={i.phase.value}")
    i.tick(200)
    print(f"after +200ms: phase={i.phase.value}")
    i.tick(100)
    print(f"after +100ms: phase={i.phase.value}")

    # Adaptive green
    i2 = Intersection(cfg)
    i2.set_ns_queue(3)
    i2.tick(1000)
    print(f"with queue: phase={i2.phase.value} duration={i2.phase_duration_ms}")

    # Emergency
    i3 = Intersection(cfg)
    i3.tick(300)
    i3.begin_emergency()
    print(f"emergency started: phase={i3.phase.value}")
    i3.tick(10000)
    print(f"frozen: phase={i3.phase.value}")
    i3.end_emergency()
    print(f"resumed: phase={i3.phase.value} elapsed={i3.phase_elapsed_ms}")

    # Pedestrian
    i4 = Intersection(cfg)
    i4.request_ped_crossing()
    print(f"ped requested, signal={i4.ped_signal().value}")
    i4.tick(1200)
    print(f"after 1200ms: phase={i4.phase.value} ped={i4.ped_signal().value}")
    i4.tick(500)
    print(f"after +500ms: ped={i4.ped_signal().value}")
    i4.tick(300)
    print(f"after +300ms: ped={i4.ped_signal().value}")


if __name__ == "__main__":
    demo()

File: rust/graphics/traffic_light.rs
//! Traffic Light — coordinated super-state FSM + preemption + adaptive timing.
//!
//! Twelfth Graphics project. Distinctive moves:
//!
//!   * **Coordinated super-state** — the intersection's state is one of
//!     five phases: NSGreen, NSYellow, AllRed, EWGreen, EWYellow. The
//!     individual light colours for each direction are *derived* from
//!     the phase. Invalid combinations (NS green while EW green) are
//!     unrepresentable.
//!   * **Emergency preemption with state preservation** — when an
//!     emergency begins, the current phase and elapsed are saved;
//!     intersection enters `EmergencyAllRed`. When emergency ends,
//!     resume from the saved phase.
//!   * **Adaptive green extension** — green phases can extend up to
//!     `max_green_extension_ms` if a queue is detected in that direction.
//!     The decision is a pure function of queue + elapsed + cap.
//!   * **Pedestrian walk gated to AllRed** — `request_ped_crossing`
//!     only triggers the walk phase during AllRed; otherwise it queues
//!     the request for the next AllRed.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Phase {
    NSGreen,
    NSYellow,
    AllRed,
    EWGreen,
    EWYellow,
    EmergencyAllRed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LightColor { Red, Yellow, Green }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PedSignal { DontWalk, Walk, Flashing }

impl Phase {
    pub fn ns_light(self) -> LightColor {
        match self {
            Phase::NSGreen => LightColor::Green,
            Phase::NSYellow => LightColor::Yellow,
            _ => LightColor::Red,
        }
    }

    pub fn ew_light(self) -> LightColor {
        match self {
            Phase::EWGreen => LightColor::Green,
            Phase::EWYellow => LightColor::Yellow,
            _ => LightColor::Red,
        }
    }

    /// The nominal next phase in the cycle (ignoring preemption).
    pub fn next_in_cycle(self) -> Phase {
        match self {
            Phase::NSGreen => Phase::NSYellow,
            Phase::NSYellow => Phase::AllRed,
            Phase::AllRed => Phase::EWGreen,
            Phase::EWGreen => Phase::EWYellow,
            Phase::EWYellow => Phase::AllRed, // then NS, but we flip after AllRed → EW → AllRed → NS via a flag
            Phase::EmergencyAllRed => Phase::EmergencyAllRed,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Config {
    pub ns_green_ms: u64,
    pub ns_yellow_ms: u64,
    pub ew_green_ms: u64,
    pub ew_yellow_ms: u64,
    pub all_red_ms: u64,
    pub max_green_extension_ms: u64,
    pub ped_walk_ms: u64,
    pub ped_flashing_ms: u64,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            ns_green_ms: 20_000,
            ns_yellow_ms: 3_000,
            ew_green_ms: 20_000,
            ew_yellow_ms: 3_000,
            all_red_ms: 2_000,
            max_green_extension_ms: 10_000,
            ped_walk_ms: 8_000,
            ped_flashing_ms: 5_000,
        }
    }
}

#[derive(Debug, Clone)]
pub enum IntersectionEvent {
    PhaseChanged { from: Phase, to: Phase, at_ms: u64 },
    GreenExtended { direction: Direction, by_ms: u64 },
    PedRequested { at_ms: u64 },
    PedWalkStarted { at_ms: u64 },
    PedWalkEnded { at_ms: u64 },
    EmergencyStarted { at_ms: u64 },
    EmergencyEnded { at_ms: u64 },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction { NS, EW }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum PedState { Inactive, Walk, Flashing }

#[derive(Debug, Clone)]
pub struct Intersection {
    pub phase: Phase,
    pub phase_elapsed_ms: u64,
    pub phase_duration_ms: u64, // effective duration (may include extension)
    pub cycle_side: Direction, // tracks AllRed → NS or AllRed → EW
    pub config: Config,
    pub ns_queue: u32,
    pub ew_queue: u32,
    pub ped_requested: bool,
    ped_state: PedState,
    ped_elapsed_ms: u64,
    saved_phase: Option<(Phase, u64, Direction)>, // (phase, elapsed, cycle_side)
    pub total_elapsed_ms: u64,
    pub events: Vec<IntersectionEvent>,
}

impl Intersection {
    pub fn new(config: Config) -> Self {
        let phase = Phase::NSGreen;
        let phase_duration_ms = config.ns_green_ms;
        Self {
            phase, phase_elapsed_ms: 0, phase_duration_ms,
            cycle_side: Direction::EW, // after next AllRed, go to EW
            config,
            ns_queue: 0, ew_queue: 0,
            ped_requested: false,
            ped_state: PedState::Inactive,
            ped_elapsed_ms: 0,
            saved_phase: None,
            total_elapsed_ms: 0,
            events: vec![],
        }
    }

    pub fn ns_light(&self) -> LightColor { self.phase.ns_light() }
    pub fn ew_light(&self) -> LightColor { self.phase.ew_light() }
    pub fn ped_signal(&self) -> PedSignal {
        match self.ped_state {
            PedState::Inactive => PedSignal::DontWalk,
            PedState::Walk => PedSignal::Walk,
            PedState::Flashing => PedSignal::Flashing,
        }
    }

    pub fn phase_duration_for(&self, p: Phase) -> u64 {
        match p {
            Phase::NSGreen => self.config.ns_green_ms,
            Phase::NSYellow => self.config.ns_yellow_ms,
            Phase::EWGreen => self.config.ew_green_ms,
            Phase::EWYellow => self.config.ew_yellow_ms,
            Phase::AllRed => self.config.all_red_ms,
            Phase::EmergencyAllRed => u64::MAX, // cleared by end_emergency()
        }
    }

    pub fn set_ns_queue(&mut self, n: u32) { self.ns_queue = n; }
    pub fn set_ew_queue(&mut self, n: u32) { self.ew_queue = n; }

    pub fn request_ped_crossing(&mut self) {
        if self.ped_state != PedState::Inactive { return; }
        if !self.ped_requested {
            self.ped_requested = true;
            self.events.push(IntersectionEvent::PedRequested {
                at_ms: self.total_elapsed_ms,
            });
        }
    }

    pub fn begin_emergency(&mut self) {
        if self.phase == Phase::EmergencyAllRed { return; }
        self.saved_phase = Some((self.phase, self.phase_elapsed_ms, self.cycle_side));
        let at = self.total_elapsed_ms;
        let from = self.phase;
        self.phase = Phase::EmergencyAllRed;
        self.phase_elapsed_ms = 0;
        self.phase_duration_ms = u64::MAX;
        self.ped_state = PedState::Inactive;
        self.events.push(IntersectionEvent::EmergencyStarted { at_ms: at });
        self.events.push(IntersectionEvent::PhaseChanged {
            from, to: Phase::EmergencyAllRed, at_ms: at,
        });
    }

    pub fn end_emergency(&mut self) {
        if self.phase != Phase::EmergencyAllRed { return; }
        let at = self.total_elapsed_ms;
        let from = self.phase;
        if let Some((saved_phase, saved_elapsed, saved_side)) = self.saved_phase.take() {
            self.phase = saved_phase;
            self.phase_elapsed_ms = saved_elapsed;
            self.phase_duration_ms = self.phase_duration_for(saved_phase);
            self.cycle_side = saved_side;
        } else {
            self.phase = Phase::AllRed;
            self.phase_elapsed_ms = 0;
            self.phase_duration_ms = self.config.all_red_ms;
        }
        self.events.push(IntersectionEvent::EmergencyEnded { at_ms: at });
        self.events.push(IntersectionEvent::PhaseChanged {
            from, to: self.phase, at_ms: at,
        });
    }

    /// Try to extend the current green phase if a queue is detected.
    /// Returns bytes extended.
    fn consider_extension(&mut self) -> u64 {
        let direction = match self.phase {
            Phase::NSGreen => Some(Direction::NS),
            Phase::EWGreen => Some(Direction::EW),
            _ => None,
        };
        let Some(dir) = direction else { return 0; };
        let queue = match dir { Direction::NS => self.ns_queue, Direction::EW => self.ew_queue };
        if queue == 0 { return 0; }

        let base = match dir {
            Direction::NS => self.config.ns_green_ms,
            Direction::EW => self.config.ew_green_ms,
        };
        let already_extended = self.phase_duration_ms.saturating_sub(base);
        let remaining_cap = self.config.max_green_extension_ms.saturating_sub(already_extended);
        if remaining_cap == 0 { return 0; }

        // grant a second batch proportional to queue (capped)
        let grant = (queue as u64 * 2_000).min(remaining_cap);
        self.phase_duration_ms += grant;
        if grant > 0 {
            self.events.push(IntersectionEvent::GreenExtended {
                direction: dir, by_ms: grant,
            });
        }
        grant
    }

    fn next_phase(&self) -> (Phase, Direction) {
        match self.phase {
            Phase::NSGreen => (Phase::NSYellow, self.cycle_side),
            Phase::NSYellow => (Phase::AllRed, Direction::EW),
            Phase::AllRed => match self.cycle_side {
                Direction::EW => (Phase::EWGreen, Direction::NS),
                Direction::NS => (Phase::NSGreen, Direction::EW),
            },
            Phase::EWGreen => (Phase::EWYellow, self.cycle_side),
            Phase::EWYellow => (Phase::AllRed, Direction::NS),
            Phase::EmergencyAllRed => (Phase::EmergencyAllRed, self.cycle_side),
        }
    }

    fn transition_phase(&mut self) -> bool {
        let (new_phase, new_side) = self.next_phase();
        let from = self.phase;
        self.phase = new_phase;
        self.cycle_side = new_side;
        self.phase_elapsed_ms = 0;
        self.phase_duration_ms = self.phase_duration_for(new_phase);
        self.events.push(IntersectionEvent::PhaseChanged {
            from, to: new_phase, at_ms: self.total_elapsed_ms,
        });

        // Gate pedestrian walk to AllRed entry
        if new_phase == Phase::AllRed && self.ped_requested {
            self.ped_requested = false;
            self.ped_state = PedState::Walk;
            self.ped_elapsed_ms = 0;
            self.events.push(IntersectionEvent::PedWalkStarted {
                at_ms: self.total_elapsed_ms,
            });
            return true;
        }
        false
    }

    fn advance_pedestrian(&mut self, ms: u64) {
        if self.ped_state == PedState::Inactive { return; }
        self.ped_elapsed_ms += ms;
        match self.ped_state {
            PedState::Walk if self.ped_elapsed_ms >= self.config.ped_walk_ms => {
                self.ped_state = PedState::Flashing;
                self.ped_elapsed_ms = 0;
            }
            PedState::Flashing if self.ped_elapsed_ms >= self.config.ped_flashing_ms => {
                self.ped_state = PedState::Inactive;
                self.ped_elapsed_ms = 0;
                self.events.push(IntersectionEvent::PedWalkEnded {
                    at_ms: self.total_elapsed_ms,
                });
            }
            _ => {}
        }
    }

    pub fn tick(&mut self, ms: u64) {
        self.total_elapsed_ms += ms;
        if self.phase == Phase::EmergencyAllRed {
            // Emergency never auto-advances; pedestrian also frozen.
            return;
        }

        self.phase_elapsed_ms += ms;

        // Try to extend green phases at the end of their nominal duration
        if (self.phase == Phase::NSGreen || self.phase == Phase::EWGreen)
            && self.phase_elapsed_ms >= self.phase_duration_ms
        {
            self.consider_extension();
        }

        let mut ped_just_started_at_remaining: Option<u64> = None;
        while self.phase_elapsed_ms >= self.phase_duration_ms
            && self.phase != Phase::EmergencyAllRed
        {
            let overflow = self.phase_elapsed_ms - self.phase_duration_ms;
            let started = self.transition_phase();
            self.phase_elapsed_ms = overflow.min(self.phase_duration_ms);
            if started {
                ped_just_started_at_remaining = Some(overflow);
            }
        }

        // Pedestrian advances by the time it has actually been active during
        // this tick: either `ms` (active before tick) or `overflow` (just started).
        let ped_advance_ms = match (self.ped_state, ped_just_started_at_remaining) {
            (PedState::Inactive, _) => 0,
            (_, Some(remaining)) => remaining,
            _ => ms,
        };
        self.advance_pedestrian(ped_advance_ms);
    }
}

// ---------- tests ----------

#[cfg(test)]
mod tests {
    use super::*;

    fn small_config() -> Config {
        Config {
            ns_green_ms: 1000,
            ns_yellow_ms: 200,
            ew_green_ms: 1000,
            ew_yellow_ms: 200,
            all_red_ms: 100,
            max_green_extension_ms: 500,
            ped_walk_ms: 500,
            ped_flashing_ms: 300,
        }
    }

    #[test]
    fn initial_phase_is_ns_green() {
        let i = Intersection::new(small_config());
        assert_eq!(i.phase, Phase::NSGreen);
        assert_eq!(i.ns_light(), LightColor::Green);
        assert_eq!(i.ew_light(), LightColor::Red);
    }

    #[test]
    fn phase_cycle_nsgreen_to_nsyellow_to_allred_to_ewgreen() {
        let mut i = Intersection::new(small_config());
        i.tick(1000); // NSGreen done
        assert_eq!(i.phase, Phase::NSYellow);
        i.tick(200); // NSYellow done
        assert_eq!(i.phase, Phase::AllRed);
        i.tick(100); // AllRed done
        assert_eq!(i.phase, Phase::EWGreen);
        i.tick(1000); // EWGreen done
        assert_eq!(i.phase, Phase::EWYellow);
        i.tick(200); // EWYellow done
        assert_eq!(i.phase, Phase::AllRed);
        i.tick(100); // AllRed done
        assert_eq!(i.phase, Phase::NSGreen);
    }

    #[test]
    fn lights_are_complementary_during_green() {
        let mut i = Intersection::new(small_config());
        assert_eq!((i.ns_light(), i.ew_light()), (LightColor::Green, LightColor::Red));
        i.tick(1000);
        assert_eq!((i.ns_light(), i.ew_light()), (LightColor::Yellow, LightColor::Red));
        i.tick(200);
        assert_eq!((i.ns_light(), i.ew_light()), (LightColor::Red, LightColor::Red));
        i.tick(100);
        assert_eq!((i.ns_light(), i.ew_light()), (LightColor::Red, LightColor::Green));
    }

    #[test]
    fn green_extends_when_queue_detected() {
        let mut i = Intersection::new(small_config());
        i.set_ns_queue(3);
        i.tick(1000); // should extend by min(3*2000, 500) = 500
        // NSGreen should still be active (not yet at duration + extension)
        assert_eq!(i.phase, Phase::NSGreen);
        assert_eq!(i.phase_duration_ms, 1500);
        i.tick(500); // extension used
        assert_eq!(i.phase, Phase::NSYellow);
    }

    #[test]
    fn extension_capped_at_max() {
        let mut i = Intersection::new(small_config());
        i.set_ns_queue(100); // huge queue would want massive extension
        i.tick(1000);
        assert_eq!(i.phase_duration_ms, 1500); // capped at 1000 + 500
    }

    #[test]
    fn no_extension_without_queue() {
        let mut i = Intersection::new(small_config());
        i.set_ns_queue(0);
        i.tick(1000);
        assert_eq!(i.phase, Phase::NSYellow);
    }

    #[test]
    fn emergency_jumps_to_emergency_all_red() {
        let mut i = Intersection::new(small_config());
        i.tick(500); // mid NSGreen
        i.begin_emergency();
        assert_eq!(i.phase, Phase::EmergencyAllRed);
        assert_eq!(i.ns_light(), LightColor::Red);
        assert_eq!(i.ew_light(), LightColor::Red);
    }

    #[test]
    fn emergency_freezes_time() {
        let mut i = Intersection::new(small_config());
        i.begin_emergency();
        i.tick(10_000);
        assert_eq!(i.phase, Phase::EmergencyAllRed);
    }

    #[test]
    fn end_emergency_restores_saved_phase() {
        let mut i = Intersection::new(small_config());
        i.tick(300); // 300ms into NSGreen
        i.begin_emergency();
        i.tick(10_000);
        i.end_emergency();
        assert_eq!(i.phase, Phase::NSGreen);
        assert_eq!(i.phase_elapsed_ms, 300); // resumed from saved
    }

    #[test]
    fn ped_request_gated_to_allred() {
        let mut i = Intersection::new(small_config());
        i.request_ped_crossing();
        assert!(i.ped_requested);
        assert_eq!(i.ped_signal(), PedSignal::DontWalk); // not activated yet
        // advance through NSGreen + NSYellow → AllRed
        i.tick(1000); // NSYellow
        i.tick(200);  // AllRed, ped walk starts
        assert_eq!(i.ped_signal(), PedSignal::Walk);
        assert!(!i.ped_requested); // consumed
    }

    #[test]
    fn ped_transitions_walk_flashing_inactive() {
        let mut i = Intersection::new(small_config());
        i.request_ped_crossing();
        i.tick(1200); // into AllRed → ped walk
        assert_eq!(i.ped_signal(), PedSignal::Walk);
        i.tick(500); // walk 500ms done → flashing
        assert_eq!(i.ped_signal(), PedSignal::Flashing);
        i.tick(300); // flashing done → inactive
        assert_eq!(i.ped_signal(), PedSignal::DontWalk);
    }

    #[test]
    fn events_log_lifecycle() {
        let mut i = Intersection::new(small_config());
        i.tick(1500);
        let phase_changes = i.events.iter().filter(|e|
            matches!(e, IntersectionEvent::PhaseChanged { .. })).count();
        assert!(phase_changes >= 2);
    }

    #[test]
    fn emergency_during_emergency_is_noop() {
        let mut i = Intersection::new(small_config());
        i.begin_emergency();
        let before = i.events.len();
        i.begin_emergency();
        assert_eq!(i.events.len(), before);
    }

    #[test]
    fn end_emergency_without_emergency_is_noop() {
        let mut i = Intersection::new(small_config());
        let before = i.events.len();
        i.end_emergency();
        assert_eq!(i.events.len(), before);
    }
}
